### PR TITLE
Add `Type::reverseArray()`

### PR DESCRIPTION
**Pull request by herndlm**
Improves return type specification of lists and various edge cases with `array_reverse()`.

### CODE REVIEW

File: src/Type/Constant/ConstantArrayType.php [hunk 1]
```diff
@@ -1118,12 +1130,21 @@ public function slice(int $offset, ?int $limit, bool $preserveKeys = false): sel
...
-			return $this->reverse(true)
-				->slice($reversedOffset, $reversedLimit, $preserveKeys)
-				->reverse(true);
+			$sliced = $reversed->slice($reversedOffset, $reversedLimit, $preserveKeys)
+				->reverseArray(TrinaryLogic::createYes());
+			if (!$sliced instanceof self) {
+				throw new ShouldNotHappenException();
+			}
+
+			return $sliced;
```
Comment (herndlm) [inline src/Type/Constant/ConstantArrayType.php]: this is a bit ugly, but didn't find a good solution for the type problems yet..

we'd need a private `reverseConstantArray()` helper to have `self` as return type most likely.

File: src/Type/Constant/ConstantArrayType.php [hunk 2]
```diff
@@ -1162,16 +1179,10 @@ public function slice(int $offset, ?int $limit, bool $preserveKeys = false): sel
 		return $preserveKeys ? $slice : $slice->reindex();
 	}
 
+	/** @deprecated Use reverseArray() instead */
```
Comment (staabm) [inline src/Type/Constant/ConstantArrayType.php:1182]: Does it make sense to deprecate the method only for the ConstantArrayType?

Comment (herndlm) [reply src/Type/Constant/ConstantArrayType.php:1182]: Sorry, I didn't fully get that. The method only exists there and nowhere else.

Comment (staabm) [reply src/Type/Constant/ConstantArrayType.php:1182]: Ohh I see ;-)

**Comment by herndlm** [comment]
this one is simple (I think) and should be ready :) sorry for the impatience, I want to work on a follow-up that should improve array_reverse for constant arrays :)

**Comment by ondrejmirtes** [comment]
Love it, thanks!

**Comment by herndlm** [comment]
thank you! felt good finally contributing something again, even if it's a simple thing like that :)

**Pull request merged**